File: find_duplicates.rb
#!/usr/bin/env ruby
# find_duplicates.rb
# encoding: UTF-8

require 'digest'
require 'find'
require 'optparse'
require 'fileutils'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

def file_hash(path, algo, partial)
  hash = case algo
         when 'md5' then Digest::MD5.new
         when 'sha1' then Digest::SHA1.new
         else raise "Unsupported algorithm"
         end
  File.open(path, 'rb') do |f|
    total = 0
    while chunk = f.read(8192)
      if partial > 0 && total + chunk.bytesize > partial
        chunk = chunk.byteslice(0, partial - total)
      end
      hash.update(chunk)
      total += chunk.bytesize
      break if partial > 0 && total >= partial
    end
  end
  hash.hexdigest
end

def find_duplicates(root, recursive, algo, partial, exclude_patterns)
  size_map = {}
  Find.find(root) do |path|
    next unless File.file?(path)
    # Исключения
    skip = false
    exclude_patterns.each do |pat|
      if File.fnmatch(pat, File.basename(path))
        skip = true
        break
      end
    end
    next if skip
    size = File.size(path)
    next if size == 0
    size_map[size] ||= []
    size_map[size] << path
  end

  duplicates = {}
  size_map.each do |size, files|
    next if files.size < 2
    hash_map = {}
    files.each do |f|
      h = file_hash(f, algo, partial)
      hash_map[h] ||= []
      hash_map[h] << f
    end
    hash_map.each do |h, group|
      next if group.size < 2
      duplicates[h] ||= []
      duplicates[h] << { size: size, files: group }
    end
  end
  duplicates
end

def format_size(size)
  units = ['B', 'KB', 'MB', 'GB', 'TB']
  s = size.to_f
  i = 0
  while s >= 1024 && i < units.length - 1
    s /= 1024
    i += 1
  end
  sprintf("%.1f %s", s, units[i])
end

def print_duplicates(duplicates, verbose)
  if duplicates.empty?
    puts colorize("Дубликатов не найдено.", :green)
    return
  end
  total_groups = 0
  total_files = 0
  total_size = 0
  duplicates.each do |h, groups|
    groups.each do |group|
      total_groups += 1
      total_files += group[:files].size
      total_size += group[:size] * (group[:files].size - 1)
      puts colorize("\nГруппа дубликатов (размер: #{format_size(group[:size])}):", :blue)
      puts "  Хеш: #{h}"
      group[:files].each_with_index do |f, i|
        if i == 0
          puts colorize("  [оригинал] #{f}", :green)
        else
          puts colorize("  [дубликат] #{f}", :yellow)
        end
      end
      puts "  Количество файлов: #{group[:files].size}" if verbose
    end
  end
  puts colorize("\nВсего групп: #{total_groups}, файлов: #{total_files}, экономия: #{format_size(total_size)}", :green)
end

options = {
  path: '.',
  recursive: true,
  algo: 'md5',
  partial: 0,
  exclude: [],
  delete: false,
  move: nil,
  hardlink: false,
  dry_run: false,
  log: nil,
  verbose: false,
  yes: false
}

OptionParser.new do |opts|
  opts.banner = "Usage: find_duplicates.rb [options] [path]"
  opts.on("-p", "--path DIR", "Path") { |v| options[:path] = v }
  opts.on("-r", "--recursive", "Recursive") { options[:recursive] = true }
  opts.on("-a", "--algo ALGO", "md5 or sha1") { |v| options[:algo] = v }
  opts.on("--partial N", Integer, "Hash only first N bytes") { |v| options[:partial] = v }
  opts.on("-e", "--exclude PAT", "Exclude pattern") { |v| options[:exclude] << v }
  opts.on("--delete", "Delete duplicates") { options[:delete] = true }
  opts.on("--move DIR", "Move duplicates to DIR") { |v| options[:move] = v }
  opts.on("--hardlink", "Replace with hardlinks") { options[:hardlink] = true }
  opts.on("-n", "--dry-run", "Dry run") { options[:dry_run] = true }
  opts.on("-l", "--log FILE", "Log file") { |v| options[:log] = v }
  opts.on("-v", "--verbose", "Verbose") { options[:verbose] = true }
  opts.on("-y", "--yes", "No confirm") { options[:yes] = true }
  opts.on("-h", "--help", "Help") { puts opts; exit }
end.parse!

root = File.expand_path(options[:path] || ARGV[0] || '.')
unless Dir.exist?(root)
  puts colorize("Ошибка: '#{root}' не является папкой", :red)
  exit 1
end

duplicates = find_duplicates(root, options[:recursive], options[:algo],
                              options[:partial], options[:exclude])
print_duplicates(duplicates, options[:verbose])

if options[:delete] || options[:move] || options[:hardlink]
  unless options[:yes]
    print colorize("Выполнить действия над дубликатами? [y/N] ", :yellow)
    ans = $stdin.gets.chomp
    unless ans.downcase == 'y'
      puts colorize("Операция отменена.", :red)
      exit
    end
  end
  duplicates.each do |h, groups|
    groups.each do |group|
      original = group[:files].first
      group[:files][1..-1].each do |dup|
        if options[:dry_run]
          puts colorize("[DRY RUN] Будет обработан дубликат: #{dup}", :yellow)
          next
        end
        if options[:delete]
          File.delete(dup)
          puts colorize("Удалён: #{dup}", :red)
        elsif options[:move]
          FileUtils.mkdir_p(options[:move])
          dest = File.join(options[:move], File.basename(dup))
          File.rename(dup, dest)
          puts colorize("Перемещён: #{dup} -> #{dest}", :yellow)
        elsif options[:hardlink]
          File.delete(dup)
          File.link(original, dup)
          puts colorize("Создана жёсткая ссылка: #{dup} -> #{original}", :blue)
        end
      end
    end
  end
end

File: find_duplicates.py
# find_duplicates.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import os
import sys
import argparse
import hashlib
import shutil
import fnmatch
from collections import defaultdict
from pathlib import Path

# ANSI colors
COLORS = {
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'reset': '\033[0m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

def get_file_hash(filepath, algo='md5', partial=0):
    """Вычисляет хеш файла. Если partial > 0, читает только первые partial байт."""
    hash_func = hashlib.new(algo)
    try:
        with open(filepath, 'rb') as f:
            if partial > 0:
                chunk = f.read(partial)
                hash_func.update(chunk)
                # Если файл больше, всё равно хешируем только начало (быстро)
            else:
                for chunk in iter(lambda: f.read(8192), b''):
                    hash_func.update(chunk)
        return hash_func.hexdigest()
    except (IOError, OSError):
        return None

def find_duplicates(root, recursive=True, algo='md5', partial=0, exclude=None):
    """Находит дубликаты файлов в указанной папке."""
    exclude = exclude or []
    size_map = defaultdict(list)
    # Первый проход: группировка по размеру
    for dirpath, dirnames, filenames in os.walk(root):
        if not recursive and dirpath != root:
            continue
        # Исключения для папок
        skip_dir = False
        for pat in exclude:
            if fnmatch.fnmatch(os.path.basename(dirpath), pat):
                skip_dir = True
                break
        if skip_dir:
            continue
        for fname in filenames:
            # Исключения для файлов
            skip_file = False
            for pat in exclude:
                if fnmatch.fnmatch(fname, pat):
                    skip_file = True
                    break
            if skip_file:
                continue
            full = os.path.join(dirpath, fname)
            try:
                size = os.path.getsize(full)
                if size > 0:  # пропускаем пустые файлы
                    size_map[size].append(full)
            except OSError:
                continue

    # Второй проход: хеширование для групп с одинаковым размером
    duplicates = defaultdict(list)
    for size, files in size_map.items():
        if len(files) < 2:
            continue
        hash_map = defaultdict(list)
        for f in files:
            h = get_file_hash(f, algo, partial)
            if h is not None:
                hash_map[h].append(f)
        for h, group in hash_map.items():
            if len(group) > 1:
                duplicates[h].append((size, group))
    return duplicates

def format_size(size):
    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
        if size < 1024.0:
            return f"{size:.1f} {unit}"
        size /= 1024.0
    return f"{size:.1f} PB"

def print_duplicates(duplicates, verbose=False):
    if not duplicates:
        print(colorize("Дубликатов не найдено.", 'green'))
        return
    total_groups = 0
    total_files = 0
    total_size = 0
    for hash_val, groups in duplicates.items():
        for size, group in groups:
            total_groups += 1
            total_files += len(group)
            total_size += size * len(group) - size  # экономия
            print(colorize(f"\nГруппа дубликатов (размер: {format_size(size)}):", 'blue'))
            print(f"  Хеш: {hash_val}")
            for i, f in enumerate(group):
                if i == 0:
                    print(colorize(f"  [оригинал] {f}", 'green'))
                else:
                    print(colorize(f"  [дубликат] {f}", 'yellow'))
            if verbose:
                print(f"  Количество файлов: {len(group)}")
    print(colorize(f"\nВсего групп: {total_groups}, файлов: {total_files}, экономия: {format_size(total_size)}", 'green'))

def main():
    parser = argparse.ArgumentParser(description="Duplicate Finder – поиск дубликатов файлов")
    parser.add_argument('path', nargs='?', default='.', help='Папка для сканирования (по умолчанию текущая)')
    parser.add_argument('-r', '--recursive', action='store_true', default=True, help='Рекурсивно (включено)')
    parser.add_argument('-a', '--algo', choices=['md5', 'sha1'], default='md5', help='Алгоритм хеширования')
    parser.add_argument('--partial', type=int, default=0, help='Хешировать только первые N байт (для больших файлов)')
    parser.add_argument('-e', '--exclude', action='append', default=[], help='Исключить по glob-шаблону')
    parser.add_argument('--delete', action='store_true', help='Удалить дубликаты (оставить по одному)')
    parser.add_argument('--move', help='Переместить дубликаты в указанную папку')
    parser.add_argument('--hardlink', action='store_true', help='Заменить дубликаты жёсткими ссылками на оригинал')
    parser.add_argument('-n', '--dry-run', action='store_true', help='Только показать действия')
    parser.add_argument('-l', '--log', help='Сохранить отчёт в файл')
    parser.add_argument('-v', '--verbose', action='store_true', help='Подробный вывод')
    parser.add_argument('-y', '--yes', action='store_true', help='Не запрашивать подтверждение')
    args = parser.parse_args()

    root = os.path.abspath(args.path)
    if not os.path.isdir(root):
        sys.exit(colorize(f"Ошибка: '{root}' не является папкой", 'red'))

    duplicates = find_duplicates(root, args.recursive, args.algo, args.partial, args.exclude)
    if args.log:
        with open(args.log, 'w') as f:
            # Сохраняем в файл
            pass  # упрощённо
    print_duplicates(duplicates, args.verbose)

    # Действия с дубликатами (если указаны)
    if args.delete or args.move or args.hardlink:
        if not args.yes:
            answer = input(colorize("Выполнить действия над дубликатами? [y/N] ", 'yellow'))
            if answer.lower() != 'y':
                print(colorize("Операция отменена.", 'red'))
                return
        # Реализация действий (упрощённо)
        for hash_val, groups in duplicates.items():
            for size, group in groups:
                original = group[0]
                for dup in group[1:]:
                    if args.dry_run:
                        print(colorize(f"[DRY RUN] Будет обработан дубликат: {dup}", 'yellow'))
                        continue
                    if args.delete:
                        os.remove(dup)
                        print(colorize(f"Удалён: {dup}", 'red'))
                    elif args.move:
                        os.makedirs(args.move, exist_ok=True)
                        dest = os.path.join(args.move, os.path.basename(dup))
                        shutil.move(dup, dest)
                        print(colorize(f"Перемещён: {dup} -> {dest}", 'yellow'))
                    elif args.hardlink:
                        os.remove(dup)
                        os.link(original, dup)
                        print(colorize(f"Создана жёсткая ссылка: {dup} -> {original}", 'blue'))

if __name__ == '__main__':
    main()
